File: test/functional/app-with-extensions/xwalk-extensions/js/echo.js
/* Copyright (c) 2014 Intel Corporation. All rights reserved.
 * Use of this source code is governed by an Apache v2 license that can be
 * found in the LICENSE-APACHE-V2 file. */
var callbacks = [];

extension.setMessageListener(function(msg) {
  // each callback only gets invoked once; we delete it after
  // calling it
  for (var i = 0; i < callbacks.length; i += 1) {
    callbacks[i](msg);
    delete callbacks[i];
  };
});

// really this should pass a JSON string which includes the ID
// of the caller; the Java code should then return this ID with its
// response, so the correct callback can be retrieved and invoked
exports.echo = function (msg, callback) {
  callbacks.push(callback);
  extension.postMessage(msg);
};

exports.echoSync = function (msg) {
  return extension.internal.sendSyncMessage(msg);
};

File: test/functional/app-with-extensions/xwalk-extensions/java/my/extensions/app/Echo.java
/* Copyright (c) 2014 Intel Corporation. All rights reserved.
 * Use of this source code is governed by an Apache v2 license that can be
 * found in the LICENSE-APACHE-V2 file. */
package my.extensions.app;

import org.xwalk.app.runtime.extension.XWalkExtensionClient;
import org.xwalk.app.runtime.extension.XWalkExtensionContextClient;

public class Echo extends XWalkExtensionClient {
  public Echo(String name, String jsApiContent, XWalkExtensionContextClient context) {
    super(name, jsApiContent, context);
  }

  @Override
  public void onMessage(int instanceId, String message) {
    String id = Integer.toString(instanceId);
    postMessage(instanceId, id + ": from java: " + message);
  }

  @Override
  public String onSyncMessage(int instanceId, String message) {
    return "From java sync: " + message;
  }
}
